ligo/src/treasuryDAO.mligo

```diff
@@ -0,0 +1,121 @@
+// SPDX-FileCopyrightText: 2020 TQ Tezos
+// SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+#include "defaults.mligo"
+#include "types.mligo"
+#include "proposal.mligo"
+#include "helper/unpack.mligo"
+#include "base_DAO.mligo"
+
+#include "treasuryDAO/types.mligo"
+
+// -------------------------------------
+// Helpers
+// -------------------------------------
+
+let unpack_transfer_type_list (key_name, p : string * ((string, bytes) map)) : transfer_type list =
+  let b = unpack_key (key_name, p)
+  in match ((Bytes.unpack b) : ((transfer_type list) option)) with
+    Some (v) -> v
+  | None -> (failwith "UNPACKING_NOT_TRANSFER_TYPE_LIST" : transfer_type list)
+
+// -------------------------------------
+// Configuration Lambdas
+// -------------------------------------
+
+let treasury_DAO_proposal_check (params, store : propose_params * storage) : bool =
+  let proposal_size = Bytes.size(Bytes.pack(params.proposal_metadata)) in
+  let frozen_scale_value = unpack_nat("a", store.extra) in
+  let frozen_extra_value = unpack_nat("b", store.extra) in
+  let max_proposal_size = unpack_nat("s_max", store.extra) in
+  let min_xtz_amount = unpack_tez("y", store.extra) in
+  let max_xtz_amount = unpack_tez("z", store.extra) in
+
+  let requred_token_lock = frozen_scale_value * proposal_size + frozen_extra_value in
+
+  if (params.frozen_token = requred_token_lock) && (proposal_size < max_proposal_size) then
+    let ts = unpack_transfer_type_list("transfers", params.proposal_metadata) in
+    let is_all_transfers_valid (is_valid, transfer_type: bool * transfer_type) =
+      if is_valid then
+        match transfer_type with
+            Token_transfer_type tt -> is_valid
+          | Xtz_transfer_type xt ->
+              if (min_xtz_amount <= xt.amount && xt.amount <= max_xtz_amount) then
+                is_valid
+              else
+                false
+      else
+        false
+    in List.fold is_all_transfers_valid ts true
+  else
+    false
+
+
+let treasury_DAO_rejected_proposal_return_value (params, store : proposal * storage) : nat =
+  let slash_scale_value = unpack_nat("c", store.extra) in
+  let slash_division_value =  unpack_nat("d", store.extra)
+  in (slash_scale_value * params.proposer_frozen_token) / slash_division_value
+
+
+let treasury_DAO_decision_lambda (proposal, store : proposal * storage) : return =
+  let propose_param : propose_params = {
+    frozen_token = proposal.proposer_frozen_token;
+    proposal_metadata = proposal.metadata
+    } in
+  let proposal_key = to_proposal_key (propose_param, proposal.proposer) in
+  let ts = unpack_transfer_type_list("transfers", proposal.metadata) in
+  let handle_transfer (acc, transfer_type : (bool * storage * operation list) * transfer_type) =
+      let (is_valid, store, ops) = acc in
```

Comment:
Here also, does it make sense to do the processing if `is_valid` is false here?

Reply:
I've updated this now.